src/Util/BaseURL.php

```diff
@@ -0,0 +1,363 @@
+<?php
+
+namespace Friendica\Util;
+
+use Friendica\Core\Config\Configuration;
+
+/**
+ * A class which checks and contains the basic
+ * environment for the BaseURL (url, urlpath, ssl_policy, hostname, scheme)
+ */
+class BaseURL
+{
+	/**
+	 * No SSL necessary
+	 */
+	const SSL_POLICY_NONE = 0;
+
+	/**
+	 * SSL is necessary
+	 */
+	const SSL_POLICY_FULL = 1;
+
+	/**
+	 * SSL is optional, but preferred
+	 */
+	const SSL_POLICY_SELFSIGN = 2;
+
+	/**
+	 * Define the Default SSL scheme
+	 */
+	const DEFAULT_SSL_SCHEME = self::SSL_POLICY_SELFSIGN;
+
+	/**
+	 * The Friendica Config
+	 * @var Configuration
+	 */
+	private $config;
+
+	/**
+	 * The server side variables
+	 * @var array
+	 */
+	private $server;
+
+	/**
+	 * The hostname of the Base URL
+	 * @var string
+	 */
+	private $hostname;
+
+	/**
+	 * The SSL_POLICY of the Base URL
+	 * @var int
+	 */
+	private $sslPolicy;
+
+	/**
+	 * The URL sub-path of the Base URL
+	 * @var string
+	 */
+	private $urlPath;
+
+	/**
+	 * The full URL
+	 * @var string
+	 */
+	private $url;
+
+	/**
+	 * The current scheme of this call
+	 * @var string
+	 */
+	private $scheme;
+
+	/**
+	 * Returns the hostname of this node
+	 * @return string
+	 */
+	public function getHostname()
+	{
+		return $this->hostname;
+	}
+
+	/**
+	 * Returns the current scheme of this call
+	 * @return string
+	 */
+	public function getScheme()
+	{
+		return $this->scheme;
+	}
+
+	/**
+	 * Returns the SSL policy of this node
+	 * @return int
+	 */
+	public function getSSLPolicy()
+	{
+		return $this->sslPolicy;
+	}
+
+	/**
+	 * Returns the sub-path of this URL
+	 * @return string
+	 */
+	public function getUrlPath()
+	{
+		return $this->urlPath;
+	}
+
+	/**
+	 * Returns the full URL of this call
+	 *
+	 * Note: $ssl parameter value doesn't directly correlate with the resulting protocol
+	 *
+	 * @param bool $ssl True, if ssl should get used
+	 *
+	 * @return string
+	 */
+	public function get($ssl = false)
+	{
+		if ($this->sslPolicy === self::SSL_POLICY_SELFSIGN && $ssl) {
+			return Network::switchScheme($this->url);
+		}
+
+		return $this->url;
+	}
+
+	/**
+	 * Save current parts of the base Url
+	 *
+	 * @param string? $hostname
+	 * @param int?    $sslPolicy
+	 * @param string? $urlPath
+	 *
+	 * @return bool true, if successful
+	 */
+	public function save($hostname = null, $sslPolicy = null, $urlPath = null)
+	{
+		$success = true;
+
+		if (!empty($hostname) && $hostname !== $this->hostname) {
+			$this->hostname  = $hostname;
+			if (!$this->config->set('config', 'hostname', $this->hostname)) {
```

Comment:
Why not do `$success = $this->config->set('config', 'hostname', $this->hostname);` instead and save anf `if`?

This goes for all three following `config->set` calls.

Reply:
Because `$success` should remain fals, even if the next setting was successful.

Reply:
It smells like this is missing exceptions. What's the point of going forward with a non-descriptive `false` if any of the three config save aren't successful?

Reply:
Hm.. The question is how can I make this step transactional. Because throwing an exception implies for me that the save wasn't successful at all.. Maybe inside a catch.. Or what do you think?

Reply:
You can call `config->save` in the `catch` block with the previous values indeed. In the current form it isn't transactional either so I'm not sure it needs to be, but if you can fit it in, that would be great. Or use `config->save` with an array of changes?

Reply:
Throwing an exception wouldn't be that easy and will lead to changing behavior in the `Config` class. Thus I don't want to change it for now.

I added a resetting logic in case one save wouldn't work.